Add example test Activity for Navigation Test Application

danesfeder opened the pull request:
This PR aims to create a more robust test application. When launching the test app now, you will be viewing `ExampleActivity`, which serves as an example of search + finding route + navigation.  It's currently pretty "watered down", but serves as a good starting point for our navigation driving custom user interfaces.

TODO:
- [x] Long press to select destination (https://github.com/mapbox/mapbox-navigation-android/pull/1328)
- [x] Logo + attribution adjustment for Bottomsheet (https://github.com/mapbox/mapbox-navigation-android/pull/1324)
- [x] Fix `ManeuverView` not showing correctly
- [x] `ExampleActivity` handle back press (https://github.com/mapbox/mapbox-navigation-android/pull/1323)

The `MainActivity` list of examples is still accessible via settings:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/8434572/46092848-c7194a00-c183-11e8-9c00-ff41ebb44f4c.gif)

tobrun commented:
One additional comment is that keyboard shows when opening the app, this is a strange UX as you can't see the edittext and when typing you can see the results coming from it. I would suggest not showing it at all on startup.

Guardiola31337 commented on app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt:
```diff
@@ -0,0 +1,143 @@
+  private val routeFinder: ExampleRouteFinder
+  private val accessToken: String = instance.resources.getString(R.string.mapbox_access_token)
+
+  init {
```
We should explore the option of adding _constructor properties_ and _factory functions_ to make `init` shorter.

danesfeder replied:
@Guardiola31337 can you provide an example of this?  I'm not entirely sure what you mean here.

Guardiola31337 replied:
I was referring to extract "file" functions to initialize the different variables and make the `init` block shorter.

Guardiola31337 commented on app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt:
```diff
@@ -0,0 +1,143 @@
+    const val ONE_SECOND_INTERVAL = 1000
+  }
+
+  val location: MutableLiveData<Location> = MutableLiveData()
```
We should explore the option of using custom `LiveData`s and `postValue` instead. Right now, because we're exposing `MutableLiveData` any other outside class could change mistakenly the values with `setValue`. The same thing applies to the rest of variables 👇

danesfeder replied:
@Guardiola31337 so like a custom class that extends from `MutableLiveData` and only exposes the `postValue` method?

Guardiola31337 replied:
`MutableLiveData` exposes [`postValue`](https://developer.android.com/reference/android/arch/lifecycle/MutableLiveData.html#postValue(T)). I was referring to using that one instead.

Guardiola31337 commented on app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleViewModel.kt:
```diff
@@ -0,0 +1,143 @@
+    })
+  }
+
+  fun onDestroy() {
```
`AndroidViewModel` has an `onCleared` method that we can override. It's called when the `Activity` is about to be destroyed so it'd remove the necessity of having to call this function from the `ExampleActivity#onDestroy` > `ExamplePresenter#onDestroy`

Guardiola31337 commented on app/src/main/java/com/mapbox/services/android/navigation/testapp/example/ui/ExampleActivity.kt:
```diff
@@ -0,0 +1,286 @@
+class ExampleActivity : AppCompatActivity(), ExampleView {
+
+  private val permissionsManager = PermissionsManager(this)
+  private lateinit var presenter: ExamplePresenter
```
What about?

```kotlin
private val presenter by lazy(mode = LazyThreadSafetyMode.NONE) {
  val viewModel = ViewModelProviders.of(this).get(ExampleViewModel::class.java)
  ExamplePresenter(this, viewModel)
}
```

This way `presenter` would be `val` (read-only) and also we'd remove the usage of `lateinit` - which we should treat carefully.

Guardiola31337 commented:
@danesfeder is this still WIP? Just need to address the comments? I'm excited to see this new 💯 example land!

danesfeder commented:
@Guardiola31337 nope, I'd say it's in a good spot to merge / iterate on in the future.  I'll look into addressing the comments today, thanks!

danesfeder commented:
@Guardiola31337 I addressed most of your feedback here, thanks!

I'd like to follow up on https://github.com/mapbox/mapbox-navigation-android/pull/1317#discussion_r221208851 and https://github.com/mapbox/mapbox-navigation-android/pull/1317#discussion_r221212052.  Imo, they aren't blocking this PR and I'd like to learn more about how to implement these comments before the changes are made.

danesfeder commented:
:shipit:
